### ISSUE TITLE
Length profile not correct

### ISSUE THREAD

**Issue by urskaufmann** [QGEP/QGEP#464]
There are some mistakes in the length profile:
If you select your length profile in two or more steps, then the green ground-line for the first manhole (most to the left) changes to the level of the first manhole in the last selection you make. And the ground-line is no more correct at the last manhole of the first selection. Example:
First selection 2 reaches
![lp1](https://user-images.githubusercontent.com/1257403/51458538-ca010180-1d55-11e9-92ee-b53eddf2d250.PNG)
Second selection: +1 reach
![lp2](https://user-images.githubusercontent.com/1257403/51458586-02a0db00-1d56-11e9-82af-c6ad1294d7d3.PNG)
Third seletction: +2 reaches
![lp3](https://user-images.githubusercontent.com/1257403/51458620-1d734f80-1d56-11e9-9c79-0527436ec257.PNG)
The correct length profile looks like that (if you select all reaches at once):
![lp1-3](https://user-images.githubusercontent.com/1257403/51458655-3ed43b80-1d56-11e9-9a6f-85853402f5bf.PNG)

An other problem: if there is a negative gradient, the length profile shows not the data in the database, but it shows a correction!! Example with pump (looks not so good):
![lp_pump](https://user-images.githubusercontent.com/1257403/51458734-86f35e00-1d56-11e9-893c-14013b46d308.PNG)
Other Example with just negative gradient:
![lp_negativgradient](https://user-images.githubusercontent.com/1257403/51458762-a1c5d280-1d56-11e9-8329-710e5df2cf22.png)
The manhole at the left side shows the deeper bottom, the reach it self shows not the negativ gradient.

**Comment by sjib** [comment]
@m-kuhn Any progress on this?

**Comment by urskaufmann** [comment]
Great - solves the first part of the issue.
Problem with the negative gradient still remains

**Comment by olivierdalang** [comment]
Yes looking into that now, sorry initially I thought it was the same issue.

**Comment by olivierdalang** [comment]
@urskaufmann I pushed another commit that should fix the second part of the issue

**Comment by urskaufmann** [comment]
Is fixed, thanks. But there are two other strange behaviours:
![capture_001_12022020_212936](https://user-images.githubusercontent.com/1257403/74375395-848bf800-4de0-11ea-9038-d542978f68dd.png)
On the left side (beginning of the profile) there is the manhole drawn till entry level=0.
If I zoom out, so I have to choose the beginning-node from the mouse-dropdownmenu, there is an entry level and the profile is ok!
On the right side (end of profile) it's the other way round: if I have to choose from the node from the dropdown-menu, there is no drawing of the last ground-line:
![capture_002_12022020_213658](https://user-images.githubusercontent.com/1257403/74375919-83a79600-4de1-11ea-9fd0-a7da729ee09e.png)

If I continue the profile and zoom in to choose the node direct, then all groundlines are ok:
![capture_003_12022020_213735](https://user-images.githubusercontent.com/1257403/74376007-a46feb80-4de1-11ea-8db9-14fa9666e3ed.png)

If I zoom out to start (with dropdown-menu) and zoom in to stop (direct select), I get the profile as it should be:
![capture_005_12022020_215135](https://user-images.githubusercontent.com/1257403/74376272-17796200-4de2-11ea-9300-517c17c5f08e.png)

**Comment by olivierdalang** [comment]
I couldn't reproduce the exact same behaviour, but found a bug with the drop down menu, which should be fixed now. Can you try again and see if this fixes the two behaviours you're mentioning ? (by the way, dropdown options should now be sorted)

**Comment by urskaufmann** [comment]
The problem with the endpoint (missing green groundline) is solved. I can not reproduce this.

At the startpoint is now always the entry level = 0 and the line goes down till 0. Does not matter how I choose the startpoint.

Other great effect from this dropdown-fix: issue 460 is fixed with this also!!

**Comment by olivierdalang** [comment]
@urskaufmann I can't reproduce this last issue. Is it possible the issues comes from the data rather than from the profile tool itself (with values of 0 or null) ?

If it comes from null values, is this something that we want managed in the profile tool (e.g. somehow show the value is unknown, using for instances dashed lines and interpolation as best guess) ?

If it does not come from the data, are you able to test on the demo data ? If so, can you tell me exactly between which nodes you get the issue, so I can reproduce the issue.

**Comment by urskaufmann** [comment]
I have in my data a lot of wastewater nodes with bottomlevel = NULL or 0.  But when I wrote this issue one year ago, I have never seen a manhole till 0.00.
With demo data it is really difficult to find examples.  
I have a strange example with manhole 4910: if the length profile is from 4910.0010 to 4910, then there is the manhole drawn to 0.00, when the length profile start with 4910 or goes from 4920 to 4910, then it looks as normal. Why this?
![case1](https://user-images.githubusercontent.com/1257403/74875767-13f65580-5363-11ea-9925-65005cb29354.JPG)
![case2](https://user-images.githubusercontent.com/1257403/74875772-1658af80-5363-11ea-9eff-6a2b49907a77.JPG)
![case3](https://user-images.githubusercontent.com/1257403/74875779-18bb0980-5363-11ea-8d65-1e6703d3bf1c.JPG)

**Comment by olivierdalang** [comment]
> I have a strange example with manhole 4910: if the length profile is from 4910.0010 to 4910, then there is the manhole drawn to 0.00, when the length profile start with 4910 or goes from 4920 to 4910, then it looks as normal. Why this?

I think it it because the reach between 4910.0010 and 4910 has 0 as `to_level`. It seems the profile "corrects" the reach in such cases by using the `from_level` instead of 0 (hence showing an horizontal reach). The manhole doesn't do that correction, hence going to 0.
Also the manhole is drawn using the reach levels, instead of the bottom_level, so that the issue doesn't appear when doing the profile of a path with correct reaches levels.

I suggest to start with removing that "correction". This wouldn't fix the profiles, which would look even worse, but at least they would be more consistent with the data.

Then, we could probably enhance the profile to use actual bottom_level from manholes, instead of the reaches levels.

And then, if null values are common (which seems to be the case in the demo data), we may want to try to display unknown heights (e.g. with a dashed line). @urskaufmann Is this something that's desired, or is it better to display data as is ?

**Comment by urskaufmann** [comment]
Sorry, I did not zoom in to controle the levels, but it is a good example anyway.
The length profile should not show nice pictures, but it should help to find mistakes in the data. Therefore it should only show data as is. Level=NULL or Level = 0 are missing values. Missing values should not be extrapolated or calculated. Example:
![case2](https://user-images.githubusercontent.com/1257403/74985755-6b1a2a00-5438-11ea-9b33-fe8d99ad2ece.JPG)

Missing cover level: dashed vertical line some meters over the reachpoint levels, no ground line. 
Missing reachpoint level: dashed horizontal line some meters, but no reach line from manhole to manhole. In the example there is no manhole at the right end of the length profile.

**Comment by urskaufmann** [comment]
![case1](https://user-images.githubusercontent.com/1257403/74986769-04960b80-543a-11ea-8617-968cb848cb18.png)

2nd example: its the same length profile I posted on 21.1.2019 with the wrong negative gradient (which is correct now). Look at the values of the start manhole: 
Bottom level: undefined. Correct, in the data is wastewaternode bottom level = NULL. 
Entry Level : 0m. Not correct. There are two connected inlets, and both have levels between cover level and outlet level.  And even if it would show one of the two levels: which one should it choose?
Better not show this value.

**Comment by m-kuhn** [comment]
It is only partially fixed.
Let's have a look at the remaining issues on March 6.

**Comment by sjib** [comment]
@urskaufmann @olivierdalang Can you update this issue with the work you have done?

**Comment by urskaufmann** [comment]
Did not look in since  3 month. Quit difficult, because there is no new version available.
Some important points are solved, but are not in the current version.
Missing: empty level-values are not handled as they should.

**Comment by urskaufmann** [comment]
Test with qgepplugin 1.5, database 1.51
Problems described last year and this winter are solved. Not solved is the problem of the missing values


Look the picture:
![lp1](https://user-images.githubusercontent.com/1257403/85626929-100a7500-b66e-11ea-8fdf-ff7183be8f9e.PNG)

Second manhole on the left with empty value in cover level: the green terrain-line is drawn direct from the first to the third manhole (why not). There is no Identifier-label for the second manhole. The grey line for the manhole go's to 0.
Empty reachpoint input level on the second manhole: the reach between the first and the second manhole is drawn horizontally from one manhole to the other. I can not see, that there is a missing value.

Second test with
![lp2](https://user-images.githubusercontent.com/1257403/85626939-17ca1980-b66e-11ea-9728-66c8d88fbd7b.PNG)

- Empty cover level and empty in and out level at the second manhole: the manhole is totally missing. No line, no label.

Proposal look the next picture:
![lp3_proposal](https://user-images.githubusercontent.com/1257403/85626994-257f9f00-b66e-11ea-82be-072c1e982f93.PNG)
